game/src/sandbox/dashboards/commuter.rs

```diff
@@ -234,11 +254,25 @@ impl State for CommuterPatterns {
             }
         }
 
-        if old_block != self.selected.as_ref().map(|b| b.id) || filter != self.filter {
+        let mut should_clear_drawables_cache = false;
+        if filter != self.filter {
             self.filter = filter;
-            if let Some(id) = self.selected.as_ref().map(|b| b.id) {
-                let block = &self.blocks[id];
+            should_clear_drawables_cache = true;
+        }
+
+        if old_base_block_id != base_block_id {
+            should_clear_drawables_cache = true;
+        }
 
+        if should_clear_drawables_cache {
+            self.clear_drawables_cache();
+        }
+
+        // Draw updates for selected block
```

Comment:
One thing I wanted to do, but it wasn't obvious to me how, was to better separate the "building the data model work" from the "doing layout" work. For example, everything before this point in this method is more "data model" oriented, and everything after is more "doing layout" (with the exception of the `count_per_block` work).
I initially thought I could move the layout work into `draw`, but per @dabreegster, we should avoid as much work as possible in the `draw` method since it's called constantly.

Reply:
Agreed about doing the split. I'd vote for just making a new helper function to do the layouting/drawing after figuring out what state we're in. Something like `fn draw_current_state(&self, base_block_id: BlockID) -> GeomBatch`. You'd call from `event` and only recalculate when the pair of base/comparison block changes, then `g.redraw` in `draw()`